File: exec/makeuser.js
writeln("\nMAKEUSER v2.0 - Adds User to Synchronet User Database\n");

(function() {
	
	/* functions */
	function openCan(filename) {
		var list = [];
		if(file_exists(filename)) {
			var f = new File(filename);
			if(f.open('r',true)) {
				list = f.readAll();
				f.close();
			}
		}
		return list;
	}
	
	function scanCan(can,text) {
		for each(var i in can) {
			pattern = i;
			if(i.match(/~$/) != null)
				pattern = "*" + i + "*";
			if(wildmatch(text,i))
				return true;
		}
		return false;
	}

	function usage() {
		writeln('usage: jsexec.exe makeuser.js name [-param value] [...]');
		writeln('');
		writeln('params:');
		writeln('');
		writeln(' -P      Password');
		writeln(' -R      Real name');
		writeln(' -H      Chat handle');
		writeln(' -G      Gender (M or F)');
		writeln(' -B      Birth date (in MM/DD/YY or DD/MM/YY format)');
		writeln(' -T      Telephone number');
		writeln(' -N      Netmail (Internet e-mail) address');
		writeln(' -A      Street address');
		writeln(' -L      Location (city, state)');
		writeln(' -Z      Zip/Postal code');
		writeln(' -S      Security level');
		writeln(' -F#     Flag set #');
		writeln(' -FE     Exemption flags');
		writeln(' -FR     Restriction flags');
		writeln(' -E      Expiration days');
		writeln(' -C      Comment');
		writeln('');
		writeln('NOTE: multi-word user name and param values must be enclosed in "quotes"');
	}
	
	/* stuff */
	var error = false;
	var badnames = openCan(system.text_dir + "name.can");
	var badpws = openCan(system.text_dir + "password.can");
	
	/* properties */
	var alias = argv.shift();
	var pass = undefined;
	var name = undefined;
	var gender = undefined;
	var birthdate = undefined;
	var telephone = undefined;
	var email = undefined;
	var address = undefined;
	var location = undefined;
	var zip = undefined;
	var level = undefined;
	var flags1 = undefined;
	var flags2 = undefined;
	var flags3 = undefined;
	var flags4 = undefined;
	var exemptions = undefined;
	var restrictions = undefined;
	var expire = undefined;
	var comment = undefined;

	if(alias == undefined) {
		usage();
		return false;
	}
	if(system.matchuser(alias) > 0) {
		writeln("* A user by that name already exists");
		return false;
	}
	if(scanCan(badnames,alias)) {
		writeln("* Illegal user alias per " + system.text_dir + "name.can");
		return false;
	}

	/* process arguments */
	while(argv.length > 0) {
		var arg = argv.shift();
		switch(arg[1].toUpperCase()) {
		case "P":
			password = argv.shift();
			if(scanCan(badpws,password)) {
				writeln("* Illegal password per " + system.text_dir + "name.can");
				error = true;
			}
			break;
		case "R":
			name = argv.shift();
			if(scanCan(badnames,name)) {
				writeln("* Illegal user name per " + system.text_dir + "name.can");
				return false;
			}
			break;
		case "H":
			handle = argv.shift().substr(0,8);
			if(scanCan(badnames,handle)) {
				writeln("* Invalid user handle per " + system.text_dir + "name.can");
				return false;
			}
			break;
		case "G":
			gender = argv.shift();
			if(gender.match(/(m|f)/i) == null) {
				writeln("* gender must be 'M' or 'F'");
				error = true;
			}
			break;
		case "B":
			birthdate = new Date(argv.shift());
			if(birthdate == "Invalid Date") {
				writeln("* Invalid birth date");
				error = true;
			}
			break;
		case "T":
			telephone = argv.shift();
			if(!telephone.match(/\d{3}-?\d{3}-?\d{4}/)) {
				writeln("* Invalid telephone number");
				error = true;
			}
			break;
		case "N":
			email = argv.shift();
			if(!wildmatch(email,"*@*.*")) {
				writeln("* Invalid e-mail address");
				error = true;
			}
			break;
		case "A":
			address = argv.shift();
			break;
		case "L":
			location = argv.shift();
			break;
		case "Z":
			zip = argv.shift();
			if(!zip.match(/\d+/)) {
				writeln("* Invalid zip code");
				error = true;
			}
			break;
		case "S":
			level = Number(argv.shift());
			if(isNaN(level)) {
				writeln("* Invalid security level");
				error = true;
			}
			break;
		case "F":
			switch(arg[2].toUpperCase()) {
			case "1":
				flags1 = flags_str(argv.shift());
				break;
			case "2":
				flags2 = flags_str(argv.shift());
				break;
			case "3":
				flags3 = flags_str(argv.shift());
				break;
			case "4":
				flags4 = flags_str(argv.shift());
				break;
			case "E":
				exemptions = flags_str(argv.shift());
				break;
			case "R":
				restrictions = flags_str(argv.shift());
				break;
			default:
				writeln("* Invalid flag set");
				error = true;
				break;
			}
			break;
		case "E":
			expire = argv.shift();
			if(isNaN(expire)) {
				writeln("* Invalid expiration days");
				error = true;
			}
			break;
		case "C":
			comment = argv.shift();
			break;
		default:
			writeln("* Invalid argument " + arg);
			error = true;
			break;
		}
	}
	
	/* if there was an error parsing arguments, exit */
	if(error) {
		writeln("\nUser record creation unsuccessful.");
		return false;
	}
	
	/* create user and assign properties */
	var newuser = system.new_user(alias);

	/* assign user properties */
	if(password)
		newuser.security.password = password;
	if(name)
		newuser.name = name;
	if(gender)
		newuser.gender = gender;
	if(handle)
		newuser.handle = handle;
	if(birthdate)
		newuser.birthdate = birthdate;
	if(telephone)
		newuser.telephone = telephone;
	if(email)
		newuser.netmail = email;
	if(address)
		newuser.address = address;
	if(location)
		newuser.location = location;
	if(zip)
		newuser.zipcode = zip;
	if(level)
		newuser.security.level = level;
	if(flags1)
		newuser.security.flags1 = flags1;
	if(flags2)
		newuser.security.flags2 = flags2;
	if(flags3)
		newuser.security.flags3 = flags3;
	if(flags4)
		newuser.security.flags4 = flags4;
	if(exemptions)
		newuser.security.exemptions = exemptions;
	if(restrictions)
		newuser.security.restrictions = restrictions;
	if(expire)
		newuser.security.expiration_date = time() + (expire*24*60*60);
	if(comment)
		newuser.comment = comment;
	
	writeln("User record #" + newuser.number + " " + newuser.alias + " created successfully.");
})();

writeln();

File: src/sbbs3/makeuser.c
/* Program to add a user to a Synchronet user database */

/****************************************************************************
 * @format.tab-size 4		(Plain Text/Source Code File Header)			*
 * @format.use-tabs true	(see http://www.synchro.net/ptsc_hdr.html)		*
 *																			*
 * Copyright Rob Swindell - http://www.synchro.net/copyright.html			*
 *																			*
 * This program is free software; you can redistribute it and/or			*
 * modify it under the terms of the GNU General Public License				*
 * as published by the Free Software Foundation; either version 2			*
 * of the License, or (at your option) any later version.					*
 * See the GNU General Public License for more details: gpl.txt or			*
 * http://www.fsf.org/copyleft/gpl.html										*
 *																			*
 * For Synchronet coding style and modification guidelines, see				*
 * http://www.synchro.net/source.html										*
 *																			*
 * Note: If this box doesn't appear square, then you need to fix your tabs.	*
 ****************************************************************************/

#include "load_cfg.h"
#include "str_util.h"
#include "userdat.h"
#include "scfglib.h"
#include <stdarg.h>

scfg_t scfg;

/****************************************************************************/
/* This is needed by load_cfg.c												*/
/****************************************************************************/
int lprintf(int level, const char *fmat, ...)
{
	va_list argptr;
	char    sbuf[512];
	int     chcount;

	va_start(argptr, fmat);
	chcount = vsprintf(sbuf, fmat, argptr);
	va_end(argptr);
	truncsp(sbuf);
	printf("%s\n", sbuf);
	return chcount;
}

char *usage = "\nusage: makeuser [ctrl_dir] name [-param value] [...]\n"
              "\nparams:\n"
              "\t-P\tPassword\n"
              "\t-R\tReal name\n"
              "\t-H\tChat handle\n"
              "\t-G\tGender (M or F)\n"
              "\t-B\tBirth date (in MM/DD/YY or DD/MM/YY format)\n"
              "\t-T\tTelephone number\n"
              "\t-N\tNetmail (Internet e-mail) address\n"
              "\t-A\tStreet address\n"
              "\t-L\tLocation (city, state)\n"
              "\t-Z\tZip/Postal code\n"
              "\t-S\tSecurity level\n"
              "\t-F#\tFlag set #\n"
              "\t-FE\tExemption flags\n"
              "\t-FR\tRestriction flags\n"
              "\t-E\tExpiration days\n"
              "\t-C\tComment\n"
              "\nNOTE: multi-word user name and param values must be enclosed in \"quotes\"\n"
;

/*********************/
/* Entry point (duh) */
/*********************/
int main(int argc, char **argv)
{
	const char* p;
	char        error[512];
	int         i;
	int         first_arg = 1;
	time_t      now;
	user_t      user;

	fprintf(stderr, "\nMAKEUSER v%s-%s - Adds User to Synchronet User Database\n"
	        , VERSION
	        , PLATFORM_DESC
	        );

	if (argc < 2) {
		printf("%s", usage);
		return 1;
	}

	if (strcspn(argv[first_arg], "/\\") != strlen(argv[first_arg]))
		p = argv[first_arg++];
	else
		p = get_ctrl_dir(/* warn: */ TRUE);

	memset(&scfg, 0, sizeof(scfg));
	scfg.size = sizeof(scfg);
	SAFECOPY(scfg.ctrl_dir, p);

	if (chdir(scfg.ctrl_dir) != 0)
		fprintf(stderr, "!ERROR changing directory to: %s", scfg.ctrl_dir);

	printf("\nLoading configuration files from %s\n", scfg.ctrl_dir);
	if (!load_cfg(&scfg, /* text: */ NULL, /* prep: */ TRUE, /* node: */ FALSE, error, sizeof(error))) {
		fprintf(stderr, "!ERROR loading configuration files: %s\n", error);
		exit(1);
	}

	now = time(NULL);

	memset(&user, 0, sizeof(user));

	/****************/
	/* Set Defaults */
	/****************/

	newuserdefaults(&scfg, &user);

	for (i = first_arg; i < argc; i++) {
		if (argv[i][0] == '-') {
			if (argv[i + 1] == NULL) {
				printf("%s", usage);
				return 1;
			}
			switch (toupper(argv[i++][1])) {
				case 'A':
					SAFECOPY(user.address, argv[i]);
					break;
				case 'B':
					SAFECOPY(user.birth, argv[i]);
					break;
				case 'L':
					SAFECOPY(user.location, argv[i]);
					break;
				case 'C':
					SAFECOPY(user.comment, argv[i]);
					break;
				case 'E':
					user.expire = (time32_t)(now + ((long)atoi(argv[i]) * 24L * 60L * 60L));
					break;
				case 'F':
					switch (toupper(argv[i - 1][2])) {
						case '1':
							user.flags1 = aftou32(argv[i]);
							break;
						case '2':
							user.flags2 = aftou32(argv[i]);
							break;
						case '3':
							user.flags3 = aftou32(argv[i]);
							break;
						case '4':
							user.flags4 = aftou32(argv[i]);
							break;
						case 'E':
							user.exempt = aftou32(argv[i]);
							break;
						case 'R':
							user.rest = aftou32(argv[i]);
							break;
						default:
							printf("%s", usage);
							return 1;
					}
					break;
				case 'G':
					user.sex = toupper(argv[i][0]);
					break;
				case 'H':
					SAFECOPY(user.handle, argv[i]);
					break;
				case 'N':
					SAFECOPY(user.netmail, argv[i]);
					break;
				case 'P':
					SAFECOPY(user.pass, argv[i]);
					strupr(user.pass);
					break;
				case 'R':
					SAFECOPY(user.name, argv[i]);
					break;
				case 'S':
					user.level = atoi(argv[i]);
					break;
				case 'T':
					SAFECOPY(user.phone, argv[i]);
					break;
				case 'Z':
					SAFECOPY(user.zipcode, argv[i]);
					break;
				default:
					printf("%s", usage);
					return 1;
			}
		}
		else
			SAFECOPY(user.alias, argv[i]);
	}

	if (user.alias[0] == 0) {
		printf("%s", usage);
		return 1;
	}

	if ((i = matchuser(&scfg, user.alias, FALSE)) != 0) {
		printf("!User (%s #%d) already exists\n", user.alias, i);
		return 2;
	}

	if (user.handle[0] == 0)
		SAFECOPY(user.handle, user.alias);
	if (user.name[0] == 0)
		SAFECOPY(user.name, user.alias);

	if ((i = newuserdat(&scfg, &user)) != 0) {
		fprintf(stderr, "!ERROR %d adding new user record\n", i);
		return i;
	}

	printf("User record #%d (%s) created successfully.\n", user.number, user.alias);

	return 0;
}

